### PR TITLE
[Misc] Add implementation of num_nodes

### PR DESCRIPTION
**Pull request by Skeleton003**
## Description

Implementation of `num_nodes` of CSCSamplingGraph.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 1]
```diff
@@ -108,6 +110,59 @@ def total_num_edges(self) -> int:
         """
         return self._c_csc_graph.num_edges()
 
+    @property
+    def num_nodes(self) -> Union[int, Dict[str, int]]:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:114]: Please add test, test should always come with the code.

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 2]
```diff
@@ -108,6 +110,59 @@ def total_num_edges(self) -> int:
...
+        - If the graph is heterogenous, returns a dictionary.
+
+        Returns
+        -------
+        Union[int, Dict[str, int]]
+            The number of nodes. Integer indicates the total nodes number of a
+            homogenous graph; dict indicates nodes number per node types of a
+            heterogenous graph.
+
+        TODO[Mingbang]: Add code examples to demonstrate the usage of num_nodes
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:126]: The code example usually should come from the test cases.

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 3]
```diff
@@ -108,6 +110,59 @@ def total_num_edges(self) -> int:
...
+        - If the graph is heterogenous, returns a dictionary.
+
+        Returns
+        -------
+        Union[int, Dict[str, int]]
+            The number of edges. Integer indicates the total edges number of a
+            homogenous graph; dict indicates edges number per edge types of a
+            heterogenous graph.
+        """
+        raise self.total_num_edges
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:164]: Just remove this should be fine. Let the parent class throw an not implemented error is better than silently provide a compromised implementation.

Comment (Skeleton003) [reply python/dgl/graphbolt/impl/csc_sampling_graph.py]:
```
python/dgl/graphbolt/impl/csc_sampling_graph.py:78:0: W0223: Method 'num_edges' is abstract in class 'SamplingGraph' but is not overridden (abstract-method)
```

message from https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-6395/6/6/logs/logs_dir/tmped_v7uvf.log

maybe I should remove this api from the parent class?

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 4]
```diff
@@ -108,6 +110,58 @@ def total_num_edges(self) -> int:
...
+            # type depends on the length of the dictionary.
+            num_nodes_per_type = {
+                _type: offset[_idx + 1] - offset[_idx]
+                for _type, _idx in self.metadata.node_type_to_id.items()
+            }
+
+            if len(num_nodes_per_type) > 1:  # True Heterogenous.
+                return num_nodes_per_type
+            else:  # Homogenous.
+                return self._c_csc_graph.num_nodes()
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:163]: It is ok to just return a dict with 1 entry.
```suggestion
            return self._c_csc_graph.num_nodes()
```

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**